### PR TITLE
numpy 1.23.0

### PR DESCRIPTION
**Pull request by h-vetinari**
Hot off the press, let's try building it.

### CODE REVIEW

**Comment by h-vetinari** [comment]
@conda-forge-admin, please rerender

**Comment by h-vetinari** [comment]
@charris @rgommers 
f2py (or its tests) now seem to need a fortran compiler. Is this intentional?
```
E          INFO: building 'callstatement' extension
E          error: extension 'callstatement' has Fortran sources but no Fortran compiler found
```

Only shows up in a handful of tests though:
```
=========================== short test summary info ============================
ERROR f2py/tests/test_return_real.py::TestCReturnReal::test_all[t4] - Runtime...
ERROR f2py/tests/test_return_real.py::TestCReturnReal::test_all[t8] - Runtime...
ERROR f2py/tests/test_return_real.py::TestCReturnReal::test_all[s4] - Runtime...
ERROR f2py/tests/test_return_real.py::TestCReturnReal::test_all[s8] - Runtime...
ERROR f2py/tests/test_semicolon_split.py::TestMultiline::test_multiline - Run...
ERROR f2py/tests/test_semicolon_split.py::TestCallstatement::test_callstatement
19897 passed, 536 skipped, 38 xfailed, 5 xpassed, 4 warnings, 6 errors in 264.34s (0:04:24)
```
It wouldn't be a problem to add a fortran compiler for the linux tests, but just wanted to double-check.

PS. Windows errors are https://github.com/conda-forge/conda-build-feedstock/pull/176

**Comment by rgommers** [comment]
> f2py (or its tests) now seem to need a fortran compiler. Is this intentional?

I haven't seen anything about that and it's not mentioned in the release notes, so I assume it's accidental. @HaoZeke the last commits on these tests are all yours, can you have a look?

**Comment by HaoZeke** [comment]
> > f2py (or its tests) now seem to need a fortran compiler. Is this intentional?
> 
> I haven't seen anything about that and it's not mentioned in the release notes, so I assume it's accidental. @HaoZeke the last commits on these tests are all yours, can you have a look?

These tests have to have a `fortran` compiler as they check the observable behavior of Fortran code in Python, so we have some options here:
- Either skip these tests
- Or have a transitive Fortran compiler dependency (in `test` perhaps) as @h-vetinari suggested

We discussed this a bit at an F2PY meeting attended by @pearu, @melissawm and me

**Comment by rgommers** [comment]
> Either skip these tests

Skip them was always the design/requirement. `f2py` tests aren't new ....

**Comment by h-vetinari** [comment]
Well, I can see @rgommer's point that it might be good to make these tests conditional on the presence of a fortran compiler (for the average numpy user), but I have nothing adding `{{ compiler("fortran") }}` to the test dependencies here - in fact, I generally try to test the maximum amount of available tests (including optional ones), so I'd be inclined to do so even if they become conditional.

**Comment by charris** [comment]
@h-vetinari Could you open a NumPy issue for this?

**Comment by h-vetinari** [comment]
> @h-vetinari Could you open a NumPy issue for this?

Done: https://github.com/numpy/numpy/issues/21621

**Comment by h-vetinari** [comment]
Aside from that the question regarding the fortran compiler, this is looking good otherwise 👍

**Comment by h-vetinari** [comment]
Good news: I tested this without a fortran compiler again (see https://github.com/conda-forge/numpy-feedstock/pull/272/commits/fe5a11870644f028d3b46625f1e23cc279ca0106), and things are passing without it. I've still added the fortran-compiler back in, because I think we should try to test also the optional parts.

From conda-forge side: ready for launch 🚀

**Comment by h-vetinari** [comment]
So https://github.com/numpy/numpy/commit/e6fe6d4efcd7a2e1e17c9e78279c6c9f848acd34 introduced a dependency on `typing_extensions`. For the moment I'm assuming this is a test-only dependency, but CC @BvB93

**Comment by rgommers** [comment]
> So [numpy/numpy@e6fe6d4](https://github.com/numpy/numpy/commit/e6fe6d4efcd7a2e1e17c9e78279c6c9f848acd34) introduced a dependency on `typing_extensions`. For the moment I'm assuming this is a test-only dependency, but CC @BvB93

Yes, it's in `test_requirements.txt` only, not in `pyproject.toml` dependency list.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -73,21 +68,51 @@ test:
...
+    - {{ compiler('c') }}        # [linux]
+    - {{ compiler('fortran') }}  # [linux]
   commands:
     - f2py -h
     # numpy.test will show SIMD features of agent (in the past, there have been
     # failures that occured depending on presence/absence of e.g. AVX512);
     # for signature of numpy.test see the following (note default: label='fast'),
     # https://github.com/numpy/numpy/blob/maintenance/1.22.x/numpy/_pytesttester.py#L81-L82
-    {% set label = "'full'" if (CI != "travis") else "'fast'" %}
+    {% set label = "'fast'" if (CI == "travis") or (aarch64 and python_impl == "pypy") else "'full'" %}
```
Comment (rgommers) [inline recipe/meta.yaml:82]: How much extra test time does this cost? I could do the "math" here, but it's not trivial to figure out on what platforms this now switches from "fast" to "full".

File: recipe/meta.yaml [hunk 2]
```diff
@@ -73,21 +68,51 @@ test:
     - pytest-xdist         # [CI != "travis"]
     - hypothesis
     - setuptools <60.0.0
-    # some linux tests need a compiler
-    - {{ compiler('c') }}  # [linux]
+    - typing_extensions
+    # some linux tests need a C compiler;
+    # extra f2py tests need a fortran compiler
```
Comment (rgommers) [inline recipe/meta.yaml:73]: This change makes sense. For the record: it was always needed for some of the optional tests. Test failures without a Fortran compiler would be a bug though - and one that I believe we fixed recently after a regression.

**Review by rgommers — COMMENTED** [review]
Overall LGTM, thanks @h-vetinari

File: recipe/meta.yaml (hunk 1 above)
Comment (h-vetinari) [reply recipe/meta.yaml:82]: It just inverted the condition for easier readability, and then added aarch+pypy because that was timing out with `'full'` also on azure.

**Comment by rgommers** [comment]
PyPy seems unhappy:
```
SafetyError: The package for pypy3.8 located at /home/conda/feedstock_root/build_artifacts/pkg_cache/pypy3.8-7.3.9-hfb6fae5_1
appears to be corrupted. The path 'bin/pypy3'
has an incorrect size.
  reported size: 6192 bytes
  actual size: 2639 bytes
```

File: recipe/meta.yaml (hunk 2 above)
Comment (rgommers) [reply recipe/meta.yaml:73]: `cffi` is fast, `mypy` tests are slow. I don't mind much either way; they're not likely to turn up new failures, so the cost-benefit here is a bit of a toss-up.

Comment (h-vetinari) [reply recipe/meta.yaml:78]: OK, I did a selection based on what I considered (not) relevant, leaving out only:
* mypy: very unlikely that the packaging could break anything about typing
* python-cov: not relevant here as we're not comparing coverage to anything
* wheel: guessing that this has no relevance for conda-forge either

I wrote "mostly optional dependencies" because now `typing_extensions` is a "hard" test requirement (for a single test). Not sure if that's intentional, but I don't mind either way.

**Comment by BvB93** [comment]
> So [numpy/numpy@e6fe6d4](https://github.com/numpy/numpy/commit/e6fe6d4efcd7a2e1e17c9e78279c6c9f848acd34) introduced a dependency on `typing_extensions`. For the moment I'm assuming this is a test-only dependency, but CC @BvB93

Can confirm, this is a test-only dependency.

**Comment by h-vetinari** [comment]
Merging this despite https://github.com/conda-forge/pypy3.6-feedstock/issues/83. Once that is fixed, I'll restart the builds.
